frontend/src/employee-frontend/components/vasu/use-vasu.ts

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: 2017-2021 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+import {
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useEffect,
+  useState
+} from 'react'
+import { isAutomatedTest } from '../../../lib-common/utils/helpers'
+import { useDebouncedCallback } from '../../../lib-common/utils/useDebouncedCallback'
+import { useRestApi } from '../../../lib-common/utils/useRestApi'
+import {
+  getVasuDocument,
+  putVasuDocument,
+  PutVasuDocumentParams,
+  VasuDocumentResponse
+} from './api'
+import { VasuContent } from './vasu-content'
+
+type State =
+  | 'loading'
+  | 'loading-error'
+  | 'clean'
+  | 'dirty'
+  | 'saving'
+  | 'save-error'
+
+export interface VasuStatus {
+  state: State
+  savedAt?: Date
+}
+
+type VasuMetadata = Omit<VasuDocumentResponse, 'content'>
+
+interface Vasu {
+  vasu: VasuMetadata | undefined
+  content: VasuContent
+  setContent: Dispatch<SetStateAction<VasuContent>>
+  status: VasuStatus
+}
+
+export function useVasu(id: string): Vasu {
+  const [status, setStatus] = useState<VasuStatus>({ state: 'loading' })
+  const [vasu, setVasu] = useState<VasuMetadata>()
+  const [content, setContent] = useState<VasuContent>({ sections: [] })
+
+  useEffect(
+    function loadVasuDocument() {
+      setStatus({ state: 'loading' })
+      void getVasuDocument(id).then((res) =>
+        res.mapAll({
+          loading: () => null,
+          failure: () => setStatus({ state: 'loading-error' }),
+          success: ({ content, ...meta }) => {
+            setStatus({ state: 'clean' })
+            setVasu(meta)
+            setContent(content)
+          }
+        })
+      )
+    },
+    [id]
+  )
+
+  const save = useRestApi(putVasuDocument, (res) =>
+    res.mapAll({
+      loading: () => null,
+      failure: () => setStatus((prev) => ({ ...prev, state: 'save-error' })),
+      success: () => setStatus({ state: 'clean', savedAt: new Date() })
+    })
+  )
+  const saveNow = useCallback(
+    (params: PutVasuDocumentParams) => {
+      setStatus((prev) => ({ ...prev, state: 'saving' }))
+      save(params)
+    },
+    [save]
+  )
```

Comment:
Just a heads-up, this `save` callback's identity changes on every render since the `setState` parameter passed to it is recreated on every render. That's why the `useCallback` used in `saveNow` does practically nothing.